nexus/db-queries/src/db/datastore/snapshot.rs

```diff
@@ -54,108 +53,109 @@ impl DataStore {
+                    // variant, so this will be returned as CollectionNotFound
+                    // due to the `insert_into` failing.
+                    //
+                    // If this function is passed a snapshot with an ID that
+                    // does not match, but a project and name that does, return
+                    // ObjectAlreadyExists here.
 
-                let existing_snapshot_id: Option<Uuid> = dsl::snapshot
-                    .filter(dsl::time_deleted.is_null())
-                    .filter(dsl::name.eq(snapshot.name().to_string()))
-                    .filter(dsl::project_id.eq(snapshot.project_id))
-                    .select(dsl::id)
-                    .limit(1)
-                    .first_async(&conn)
-                    .await
-                    .optional()?;
+                    let existing_snapshot_id: Option<Uuid> = dsl::snapshot
+                        .filter(dsl::time_deleted.is_null())
+                        .filter(dsl::name.eq(snapshot.name().to_string()))
+                        .filter(dsl::project_id.eq(snapshot.project_id))
+                        .select(dsl::id)
+                        .limit(1)
+                        .first_async(&conn)
+                        .await
+                        .optional()?;
 
-                if let Some(existing_snapshot_id) = existing_snapshot_id {
-                    if existing_snapshot_id != snapshot.id() {
-                        return Err(TransactionError::CustomError(
-                            CustomError::ResourceAlreadyExists,
-                        ));
-                    }
-                }
-
-                Project::insert_resource(
-                    project_id,
-                    diesel::insert_into(dsl::snapshot)
-                        .values(snapshot)
-                        .on_conflict((dsl::project_id, dsl::name))
-                        .filter_target(dsl::time_deleted.is_null())
-                        .do_update()
-                        .set(dsl::time_modified.eq(dsl::time_modified)),
-                )
-                .insert_and_get_result_async(&conn)
-                .await
-                .map_err(|e| {
-                    TransactionError::CustomError(CustomError::InsertError(e))
-                })
-            })
-            .await
-            .map_err(|e: TxnError| match e {
-                TxnError::CustomError(e) => match e {
-                    CustomError::ResourceAlreadyExists => {
-                        Error::ObjectAlreadyExists {
-                            type_name: ResourceType::Snapshot,
-                            object_name: snapshot_name,
+                    if let Some(existing_snapshot_id) = existing_snapshot_id {
+                        if existing_snapshot_id != snapshot.id() {
+                            return Err(
+                                err.bail(CustomError::ResourceAlreadyExists)
```

Comment:
I think you can set `Error::ObjectAlreadyExists` here and just drop `CustomError` altogether

Reply:
Good catch, updated!